### DISCUSSION TITLE
关于RemotingClient异步请求后钩子函数的执行问题

### DISCUSSION THREAD

**Discussion by FantasyChen-DH** [apache/rocketmq#6323] [Q&A] [answered]
在我最近阅读remoting包下的NettyRemotingClient和NettyRemotingAbstract时，发现在进行异步请求后，并不会对RPCHook进行进一步调用。我对此还使用test包下的RemotingServerTest进行测试：
`  @Test
    public void testInvokeAsync() throws InterruptedException, RemotingConnectException,
        RemotingTimeoutException, RemotingTooMuchRequestException, RemotingSendRequestException {

        final CountDownLatch latch = new CountDownLatch(1);
        RPCHook rpcHook = new RPCHook() {
            @Override
            public void doBeforeRequest(String remoteAddr, RemotingCommand request) {
                System.out.println("send Request");
            }

            @Override
            public void doAfterResponse(String remoteAddr, RemotingCommand request,
                                        RemotingCommand response) {
                System.out.println("after response");
            }
        };
        remotingClient.registerRPCHook(rpcHook);
        RemotingCommand request = RemotingCommand.createRequestCommand(0, null);
        request.setRemark("messi");
        remotingClient.invokeAsync("localhost:8888", request, 1000 * 3, new InvokeCallback() {
            @Override
            public void operationComplete(ResponseFuture responseFuture) {
                assertTrue(responseFuture != null);
//                assertThat(responseFuture.getResponseCommand().getLanguage()).isEqualTo(LanguageCode.JAVA);
//                assertThat(responseFuture.getResponseCommand().getExtFields()).hasSize(2);
                System.out.println("收到回应");
                latch.countDown();
            }
        });
        latch.await();

    }
}`

但未执行RPCHook中的doAfterResponse方法，所以对此有疑惑，这个钩子函数会在何时才会执行还是说异步请求无需进行，感谢解答！

**Comment by echooymxq** [comment]
这里好像有一个相关issue https://github.com/apache/rocketmq/issues/1709, /cc @aaron-ai

**Reply by FantasyChen-DH** [reply]
看样子是的 但是并没有回应 是异步请求不需要做钩子函数的After操作吗

**Comment by aaron-ai** [comment, marked as answer]
@FantasyChen-DH I think it is a bug rather than a feature, the execution of hook should not depend on the sync/async way of the invocation. It will be great if you could fix it.

**Reply by FantasyChen-DH** [reply]
OK I would like to fix it
